### ISSUE TITLE
Microsoft.Data.SqlClient: Microsoft.Data.SqlClient is not supported on this platform.

### ISSUE THREAD

**Issue by mswilson4040** [Azure/Azure-Functions#1370]
I am migrating from a functions 2.0 project to a 3.0 project. I followed the steps outlined [here](https://dev.to/azure/develop-azure-functions-using-net-core-3-0-gcm) but am unable to run the project (fails at runtime)

Error:

> Microsoft.Data.SqlClient: Microsoft.Data.SqlClient is not supported on this platform.

Here is my package dependencies:
```
    <PackageReference Include="CsvHelper" Version="12.1.2" />
    <PackageReference Include="Microsoft.Azure.Functions.Extensions" Version="1.0.0" />
    <PackageReference Include="Microsoft.Azure.WebJobs.Extensions.Storage" Version="3.0.10" />
    <PackageReference Include="Microsoft.NET.Sdk.Functions" Version="1.0.30-beta2" />
    <PackageReference Include="Octokit" Version="0.36.0" />
```

However, I suspect that this error is coming from my Database Project which is using Entity Framework 3.0. Here's those dependencies:
```
      <PackageReference Include="CsvHelper" Version="12.1.2" />
      <PackageReference Include="EFCore.BulkExtensions" Version="3.0.0" />
      <PackageReference Include="Microsoft.AspNetCore.Identity" Version="2.2.0" />
      <PackageReference Include="Microsoft.AspNetCore.Identity.EntityFrameworkCore" Version="3.0.0" />
      <PackageReference Include="Microsoft.EntityFrameworkCore" Version="3.0.0" />
      <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="3.0.0" />
      <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="3.0.0" />
```

I am not explicitly pulling in `Microsoft.Data.SqlClient` anywhere within my project / solution. I did find the reference in the Entity Framework packages though, so I assume it's coming from that.

The only other thing worth mentioning here is that my Web App (running dotnet 3.0) is able to utilize the same database project without any issues.

**dotnet --info**

```
.NET Core SDK (reflecting any global.json):
 Version:   3.0.100
 Commit:    04339c3a26

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  10.15
 OS Platform: Darwin
 RID:         osx.10.15-x64
 Base Path:   /usr/local/share/dotnet/sdk/3.0.100/

Host (useful for support):
  Version: 3.0.0
  Commit:  95a0a61858

.NET Core SDKs installed:
  2.1.4 [/usr/local/share/dotnet/sdk]
  2.1.401 [/usr/local/share/dotnet/sdk]
  2.2.106 [/usr/local/share/dotnet/sdk]
  3.0.100 [/usr/local/share/dotnet/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.3 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.4 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.3 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.4 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.0.5 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.3 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.13 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.4 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]

```

**Comment by mswilson4040** [comment]
I don't know if this is the reason or not, but there https://github.com/dotnet/sqlclient docs suggest 3.0 isn't supported?

> Supportability
> The Microsoft.Data.SqlClient package supports the below environments:
> 
> .NET Framework 4.6+
> .NET Core 2.1+
> .NET Standard 2.0+.
> The source code of this library is now available under the MIT license.

**Comment by smairo** [comment]
We have this same exact problem

**Comment by ngg** [comment]
I had this problem on .NET Core 2.1, Functions ~2, when I tried to update to Microsoft.Data.SqlClient from System.Data.SqlClient. It worked when I tried locally using func on Linux, but failed with this exception in the cloud.

**Comment by mswilson4040** [comment]
@ngg That's the thing. I haven't imported that anywhere. It's packaged with Entity Framework from what I can tell. Upgrading/downgrading that package isn't an option.

**Comment by paulbatum** [comment]
I think I know what is going on here. Try adding the following to your .csproj:

```
<Target Name="PostBuild" AfterTargets="PostBuildEvent">
    <Exec Command="copy $(OutDir)$(ProjectName).deps.json $(OutDir)bin\function.deps.json" />
  </Target>
  <Target Name="PostPublish" BeforeTargets="Publish">
    <Exec Command="copy $(PublishDir)$(ProjectName).deps.json $(PublishDir)bin\function.deps.json" />
  </Target>
```

(in the future this workaround should not be necessary)

**Comment by mswilson4040** [comment]
@paulbatum Thanks.

For anyone else having this issue and working on a mac, I had to modify what @paulbatum provided a little bit:


```
  <Target Name="PostBuild" AfterTargets="PostBuildEvent">
    <Exec Command="**cp** $(OutDir)**Chadwick.**$(ProjectName).deps.json $(OutDir)**/**bin/function.deps.json" />
  </Target>
  <Target Name="PostPublish" BeforeTargets="Publish">
    <Exec Command="**cp** $(PublishDir)**Chadwick.**$(ProjectName).deps.json $(PublishDir)**/**bin/function.deps.json" />
  </Target>
```

$(ProjectName) wasn't correct for me. It wanted the namespace I guess? In any case, aside from windows `copy` command needing to be `cp` on a mac along with minor path updates, this workaround unblocks me. Thanks, @paulbatum

**Comment by kaeus** [comment]
@paulbatum I've attempted to use this solution with no luck. I've verified that when deployed the functions.deps.json is present and accurate, however I still get the same exception as above.

I have tried a few setups to verify that the issue happens regardless of if my Azure Function project directly references Microsoft.Data.SqlClient or not. Any other suggestions if this workaround doesn't solve the issue?

**Comment by IGx89** [comment]
@kaeus you're probably hitting https://github.com/Azure/azure-functions-vs-build-sdk/issues/333. You're also need to add the following to work around that .NET Core 3.0 SDK bug:

```
  <Target Name="PostPublish" AfterTargets="AfterPublish">
    <Exec Command="move $(PublishDir)\runtimes $(PublishDir)\bin" />
  </Target>
```

@paulbatum that's awesome that there's hope for that not being needed in the future. Do you have an issue, version, and/or ETA for the fix? There's a long (closed but active) issue, https://github.com/Azure/azure-functions-host/issues/3568, with lots of people reporting the issue but no updates from Microsoft so it was seeming this wasn't ever going to be fixed.

**Comment by kaeus** [comment]
@IGx89 Thanks, that seems to have wrapped up the issue for me (combining that with the previous solutions)!

**Comment by iRubens** [comment]
For anyone hitting this problem on "zip deployed" azure functions the final .csproject section required is the following:
```
<Target Name="PostBuild" AfterTargets="PostBuildEvent">
    <Exec Command="copy $(OutDir)$(ProjectName).deps.json $(OutDir)bin\function.deps.json" />
  </Target>
  <Target Name="PostPublish" BeforeTargets="Publish">
    <Exec Command="copy $(PublishDir)$(ProjectName).deps.json $(PublishDir)bin\function.deps.json" />
    <Exec Command="move $(PublishDir)\runtimes $(PublishDir)\bin" />
  </Target>
```
@IGx89 your suggestion fixed the problem, but with "zip deployed" azure functions the directory move was being done after the publishing.

**Comment by IGx89** [comment]
@iRubens all my functions are zip deployed as well but the zipping is done as a separate command after `dotnet publish`. It sounds like you know of a way to get the dotnet CLI to zip up the publish folder for you as well? If so I'd love to know how to do that, I've had zero luck finding out how to do that myself.

**Comment by kaeus** [comment]
> @iRubens all my functions are zip deployed as well but the zipping is done as a separate command after `dotnet publish`, right before I POST the zip to Azure. It sounds like you know of a way to get the dotnet CLI to zip up the publish folder for you as well? If so I'd love to know how to do that, I've had zero luck finding out how to do that myself.

If you are using Azure DevOps to publish with the DotNetCoreCLI task you can set zipAfterPublish to true that should do that automatically. I don't know of any way to do that via `dotnet publish` alone without an extra step.

**Comment by iRubens** [comment]
@IGx89 at the moment we're deploying our functions directly from Visual Studio (I've updated my comment). Unfortunately don't know either if there's any command for the CLI to obtain the same result.

**Comment by brettsam** [comment]
For those of you working around this -- the deps.json file should now be auto-copied when using Microsoft.NET.Sdk.Functions version 3.0.0-preview1. 

There is still a problem with having the runtimes folder moved to under the bin folder during publish -- I'm looking into that now, but the workaround above (https://github.com/Azure/Azure-Functions/issues/1370#issuecomment-546030332) should work.

**Comment by shailendrarampal** [comment]
> For anyone hitting this problem on "zip deployed" (from Visual Studio) azure functions the final .csproject section required is the following:
> 
> ```
> <Target Name="PostBuild" AfterTargets="PostBuildEvent">
>   <Exec Command="copy $(OutDir)$(ProjectName).deps.json $(OutDir)bin\function.deps.json" />
> </Target>
> <Target Name="PostPublish" BeforeTargets="Publish">
>   <Exec Command="copy $(PublishDir)$(ProjectName).deps.json $(PublishDir)bin\function.deps.json" />
>   <Exec Command="move $(PublishDir)\runtimes $(PublishDir)\bin" />
> </Target>
> ```
> 
> @IGx89 your suggestion fixed the problem, but with "zip deployed" azure functions the directory move was being done after the publishing.

My azure function stooped pubishing after adding this : https://stackoverflow.com/questions/58833050/azure-functions-runtime-exception-the-type-initializer-for-system-data-sqlclien

**Comment by brettsam** [comment]
@shailendrarampal -- what OS are you running on? Note that you should *not* need the deps.json lines anymore when using 3.0.0-preview1.

**Comment by shailendrarampal** [comment]
> @shailendrarampal -- what OS are you running on? Note that you should _not_ need the deps.json lines anymore when using 3.0.0-preview1.

I am using windows 10 64bit!
But if you go through the link above, my real problem was 'sni.dll was not loded' with aspnet core 3.0 and azure function v3-preview with system.data.sqlclient version 4.7.0

**Comment by brettsam** [comment]
This line should be taking care of that:   `<Exec Command="move $(PublishDir)\runtimes $(PublishDir)\bin" />` Your stackoverflow post doesn't show that as included, so I'm not sure exactly what you're running -- do you have that in your csproj? Can you share your entire csproj?

**Comment by shailendrarampal** [comment]
> This line should be taking care of that: `<Exec Command="move $(PublishDir)\runtimes $(PublishDir)\bin" />` Your stackoverflow post doesn't show that as included, so I'm not sure exactly what you're running -- do you have that in your csproj? Can you share your entire csproj?

`<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp3.0</TargetFramework>
    <AzureFunctionsVersion>v3-preview</AzureFunctionsVersion>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.ApplicationInsights.AspNetCore" Version="2.8.2" />
    <PackageReference Include="Microsoft.Data.SqlClient" Version="1.0.19269.1" />
    <PackageReference Include="Microsoft.NET.Sdk.Functions" Version="1.0.30-beta2" />
    <PackageReference Include="System.Data.SqlClient" Version="4.7.0" />
  </ItemGroup>
  <ItemGroup>
    <None Update="host.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
    <None Update="local.settings.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <CopyToPublishDirectory>Never</CopyToPublishDirectory>
    </None>
  </ItemGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.ApplicationInsights.AspNetCore" Version="2.8.0" />
  </ItemGroup>

  <ItemGroup>
    <None Include="$(USERPROFILE)\.nuget\packages\\microsoft.data.sqlclient\1.0.19249.1\runtimes\win\lib\netcoreapp2.1\microsoft.Data.SqlClient.dll">
      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
    </None>
  </ItemGroup>
  <Target Name="CopyToBin" BeforeTargets="Build">
    <Copy SourceFiles="$(USERPROFILE)\.nuget\packages\microsoft.data.sqlclient\1.0.19249.1\runtimes\win\lib\netcoreapp2.1\microsoft.Data.SqlClient.dll" DestinationFolder="$(OutputPath)\bin" />
  </Target>

  <Target Name="PostBuild" AfterTargets="PostBuildEvent">
    <Exec Command="copy $(OutDir)$(ProjectName).deps.json $(OutDir)bin\function.deps.json" />
  </Target>
  <Target Name="PostPublish" BeforeTargets="Publish">
    <Exec Command="copy $(PublishDir)$(ProjectName).deps.json $(PublishDir)bin\function.deps.json" />
    <Exec Command="move $(PublishDir)\runtimes $(PublishDir)\bin" />
  </Target>
  
</Project>`

**Comment by shailendrarampal** [comment]
> This line should be taking care of that: `<Exec Command="move $(PublishDir)\runtimes $(PublishDir)\bin" />` Your stackoverflow post doesn't show that as included, so I'm not sure exactly what you're running -- do you have that in your csproj? Can you share your entire csproj?

`
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp3.0</TargetFramework>
    <AzureFunctionsVersion>v3-preview</AzureFunctionsVersion>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.ApplicationInsights.AspNetCore" Version="2.8.2" />
    <PackageReference Include="Microsoft.Data.SqlClient" Version="1.0.19269.1" />
    <PackageReference Include="Microsoft.NET.Sdk.Functions" Version="1.0.30-beta2" />
    <PackageReference Include="System.Data.SqlClient" Version="4.7.0" />
  </ItemGroup>
  <ItemGroup>
    <None Update="host.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
    <None Update="local.settings.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <CopyToPublishDirectory>Never</CopyToPublishDirectory>
    </None>
  </ItemGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.ApplicationInsights.AspNetCore" Version="2.8.0" />
  </ItemGroup>

  <ItemGroup>
    <None Include="$(USERPROFILE)\.nuget\packages\\microsoft.data.sqlclient\1.0.19249.1\runtimes\win\lib\netcoreapp2.1\microsoft.Data.SqlClient.dll">
      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
    </None>
  </ItemGroup>
  <Target Name="CopyToBin" BeforeTargets="Build">
    <Copy SourceFiles="$(USERPROFILE)\.nuget\packages\microsoft.data.sqlclient\1.0.19249.1\runtimes\win\lib\netcoreapp2.1\microsoft.Data.SqlClient.dll" DestinationFolder="$(OutputPath)\bin" />
  </Target>

  <Target Name="PostBuild" AfterTargets="PostBuildEvent">
    <Exec Command="copy $(OutDir)$(ProjectName).deps.json $(OutDir)bin\function.deps.json" />
  </Target>
  <Target Name="PostPublish" BeforeTargets="Publish">
    <Exec Command="copy $(PublishDir)$(ProjectName).deps.json $(PublishDir)bin\function.deps.json" />
    <Exec Command="move $(PublishDir)\runtimes $(PublishDir)\bin" />
  </Target>
  
</Project>
`

**Comment by shailendrarampal** [comment]
```<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp3.0</TargetFramework>
    <AzureFunctionsVersion>v3-preview</AzureFunctionsVersion>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.ApplicationInsights.AspNetCore" Version="2.8.2" />
    <PackageReference Include="Microsoft.Data.SqlClient" Version="1.0.19269.1" />
    <PackageReference Include="Microsoft.NET.Sdk.Functions" Version="1.0.30-beta2" />
    <PackageReference Include="System.Data.SqlClient" Version="4.7.0" />
  </ItemGroup>
  <ItemGroup>
    <None Update="host.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
    <None Update="local.settings.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <CopyToPublishDirectory>Never</CopyToPublishDirectory>
    </None>
  </ItemGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.ApplicationInsights.AspNetCore" Version="2.8.0" />
  </ItemGroup>

  <ItemGroup>
    <None Include="$(USERPROFILE)\.nuget\packages\\microsoft.data.sqlclient\1.0.19249.1\runtimes\win\lib\netcoreapp2.1\microsoft.Data.SqlClient.dll">
      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
    </None>
  </ItemGroup>
  <Target Name="CopyToBin" BeforeTargets="Build">
    <Copy SourceFiles="$(USERPROFILE)\.nuget\packages\microsoft.data.sqlclient\1.0.19249.1\runtimes\win\lib\netcoreapp2.1\microsoft.Data.SqlClient.dll" DestinationFolder="$(OutputPath)\bin" />
  </Target>

  <Target Name="PostBuild" AfterTargets="PostBuildEvent">
    <Exec Command="copy $(OutDir)$(ProjectName).deps.json $(OutDir)bin\function.deps.json" />
  </Target>
  <Target Name="PostPublish" BeforeTargets="Publish">
    <Exec Command="copy $(PublishDir)$(ProjectName).deps.json $(PublishDir)bin\function.deps.json" />
    <Exec Command="move $(PublishDir)\runtimes $(PublishDir)\bin" />
  </Target>
  
</Project>```

**Comment by shailendrarampal** [comment]
> This line should be taking care of that: `<Exec Command="move $(PublishDir)\runtimes $(PublishDir)\bin" />` Your stackoverflow post doesn't show that as included, so I'm not sure exactly what you're running -- do you have that in your csproj? Can you share your entire csproj?

```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp3.0</TargetFramework>
    <AzureFunctionsVersion>v3-preview</AzureFunctionsVersion>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.ApplicationInsights.AspNetCore" Version="2.8.2" />
    <PackageReference Include="Microsoft.Data.SqlClient" Version="1.0.19269.1" />
    <PackageReference Include="Microsoft.NET.Sdk.Functions" Version="1.0.30-beta2" />
    <PackageReference Include="System.Data.SqlClient" Version="4.7.0" />
  </ItemGroup>
  <ItemGroup>
    <None Update="host.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
    <None Update="local.settings.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <CopyToPublishDirectory>Never</CopyToPublishDirectory>
    </None>
  </ItemGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.ApplicationInsights.AspNetCore" Version="2.8.0" />
  </ItemGroup>

  <ItemGroup>
    <None Include="$(USERPROFILE)\.nuget\packages\\microsoft.data.sqlclient\1.0.19249.1\runtimes\win\lib\netcoreapp2.1\microsoft.Data.SqlClient.dll">
      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
    </None>
  </ItemGroup>
  <Target Name="CopyToBin" BeforeTargets="Build">
    <Copy SourceFiles="$(USERPROFILE)\.nuget\packages\microsoft.data.sqlclient\1.0.19249.1\runtimes\win\lib\netcoreapp2.1\microsoft.Data.SqlClient.dll" DestinationFolder="$(OutputPath)\bin" />
  </Target>

  <Target Name="PostBuild" AfterTargets="PostBuildEvent">
    <Exec Command="copy $(OutDir)$(ProjectName).deps.json $(OutDir)bin\function.deps.json" />
  </Target>
  <Target Name="PostPublish" BeforeTargets="Publish">
    <Exec Command="copy $(PublishDir)$(ProjectName).deps.json $(PublishDir)bin\function.deps.json" />
    <Exec Command="move $(PublishDir)\runtimes $(PublishDir)\bin" />
  </Target>
  
</Project>
```

**Comment by shailendrarampal** [comment]
> This line should be taking care of that: `<Exec Command="move $(PublishDir)\runtimes $(PublishDir)\bin" />` Your stackoverflow post doesn't show that as included, so I'm not sure exactly what you're running -- do you have that in your csproj? Can you share your entire csproj?

It worked for me by adding below code in csproj file :
```
<Target Name="PostPublish" BeforeTargets="Publish">
  <Exec Command="move $(PublishDir)\runtimes $(PublishDir)\bin" />
</Target>
```
Thank you.

**Comment by brettsam** [comment]
I've just pushed Microsoft.Net.Sdk.Functions 3.0.0-preview2. I believe this package now handles all the problems mentioned in this issue. If you could try removing the workarounds and using that package, that'd be helpful. I'm going to close this now, but let me know if something hasn't been fixed and I can re-open (or create a separate issue).

**Comment by IGx89** [comment]
I can confirm both issues are fixed now for me (under Functions v3-preview, .NET Core SDK 3.0), thanks a lot @brettsam!

**Comment by zpertee** [comment]
Should this be working in Visual Studio 2019 Preview?  I was able to create the v3-preview function/.net core 3.0 SDK, but I'm still getting the not supported on this platform error message.

**Comment by brettsam** [comment]
Also make sure you're on Microsoft.Net.Sdk.Functions 3.0.0-preview2. I don't believe our templates create projects with this reference yet.

**Comment by Robzilla** [comment]
I am still getting this error with a fresh functions project in Microsoft Visual Studio Enterprise 2019 Preview Version 16.4.0 Preview 6.0.

Project file is as follows

<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp3.0</TargetFramework>
    <AzureFunctionsVersion>v3-preview</AzureFunctionsVersion>
    <AssemblyName>MyProj.Core</AssemblyName>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.Azure.Cosmos.Table" Version="1.0.5" />
    <PackageReference Include="Microsoft.Azure.Functions.Extensions" Version="1.0.0" />
    <PackageReference Include="Microsoft.Data.SqlClient" Version="1.1.0" />
    <PackageReference Include="Microsoft.NET.Sdk.Functions" Version="3.0.0-preview2" />
    <PackageReference Include="Microsoft.NETCore.App" Version="2.2.8" />
  </ItemGroup>
  <ItemGroup>
    <None Update="host.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
    <None Update="local.settings.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <CopyToPublishDirectory>Never</CopyToPublishDirectory>
    </None>
  </ItemGroup>
</Project>

**Comment by apapapapapa** [comment]
I am also still seeing the same issue,

I'm running VS preview 6.0 
see below for the csproj:

<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp3.0</TargetFramework>
    <AzureFunctionsVersion>v3-preview</AzureFunctionsVersion>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.Azure.WebJobs.Extensions.Storage" Version="3.0.4" />
    <PackageReference Include="Microsoft.Data.SqlClient" Version="1.1.0" />
    <PackageReference Include="Microsoft.NET.Sdk.Functions" Version="3.0.0-preview2" />
    <PackageReference Include="System.Data.SqlClient" Version="4.7.0" />
    <PackageReference Include="System.Text.Json" Version="4.6.0" />
  </ItemGroup>
  <ItemGroup>
    <None Update="host.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
    <None Update="local.settings.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <CopyToPublishDirectory>Never</CopyToPublishDirectory>
    </None>
  </ItemGroup>
</Project>

the function.deps.json does indeed get created into the bin folder properly with the preview2 version of Microsoft.NET.Sdk.Functions

**Comment by zpertee** [comment]
@apapapapapa Can you confirm that you've set the AzureFunctionsHiddenTagsVisible environment variable and pulled down the v3-preview templates (https://dev.to/azure/develop-azure-functions-using-net-core-3-0-gcm)?

**Comment by apapapapapa** [comment]
Ok I think I did set the ENV var and have the sdk installed but just to make sure, I started from scratch again. made sure to download the .NET Core 3 sdk, created a new function project, see below I pick v3 preview ( I assume this answers your question @zpertee )
![image](https://user-images.githubusercontent.com/47064141/69890989-0a9ea500-12ae-11ea-8fbd-142b15418c38.png)

i just changed the string in the blob trigger to point at my table.
run it locally, works fine.

then,

- use nuget to install System.Data.SqlClient v4.7.0 
- add using System.Data.SqlClient;
- add SqlConnection connection = new SqlConnection(); into the Run of my Function class
- build, run and I get this:

![image](https://user-images.githubusercontent.com/47064141/69891148-2ce4f280-12af-11ea-94e8-de70be961cdc.png)

**Comment by zpertee** [comment]
Great!  Looks like you're definitely on the right track (or at least the track I took myself).  What I discovered is that I could not use System.Data.SqlClient v4.7.0, but was able to use Microsoft.Data.SqlClient 1.1.0.   It appears that this is new library to use for SQL access: https://devblogs.microsoft.com/dotnet/introducing-the-new-microsoftdatasqlclient/.  These two libraries are extremely similar and you should be able to swap out to Microsoft.Data.SqlClient with very minimal (if any) code changes.

**Comment by apapapapapa** [comment]
oh right, I was confused about this one too, I tried your suggestion, and now I get the same but with Microsoft.Data... "System.PlatformNotSupportedException: 'Microsoft.Data.SqlClient is not supported on this platform.'"

**Comment by zpertee** [comment]
@apapapapapa Ok.  I will have to leave this one to Microsoft (@brettsam).  Sorry.  Figured I'd at least share my two cents of what worked for me.  Not worth much, I know...

**Comment by brettsam** [comment]
I'm able to reproduce this with the `System.Data.SqlClient` nuget package (investigating it now) -- but when I switch to `Microsoft.Data.SqlClient`, it starts to work.

@apapapapapa -- is that your latest project file above, or have you tweaked it since then? I see you're using both `Microsoft.Data.SqlClient` and `System.Data.SqlClient` there, which may be the problem...

**Comment by brettsam** [comment]
Okay I think I know why `System.Data.SqlClient` isn't working. I've filed https://github.com/Azure/azure-functions-host/issues/5315 to fix this in v3.

If you need to use this, you'll need to reference 4.3.1 for now -- but I think that swapping to `Microsoft.Data.SqlClient` should work for most cases.

**Comment by apapapapapa** [comment]
Ok, this is odd. Either something changed or I did have the reference to the System.Data.SqlClient in there somehow. 
Today I tried again, and it seems to be all working now.
For reference I am using the following csproj successfully now:
```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp3.0</TargetFramework>
    <AzureFunctionsVersion>v3-preview</AzureFunctionsVersion>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.Azure.WebJobs.Extensions.Storage" Version="3.0.4" />
    <PackageReference Include="Microsoft.Data.SqlClient" Version="1.1.0" />
    <PackageReference Include="Microsoft.NET.Sdk.Functions" Version="3.0.0-preview2" />
    <PackageReference Include="System.Text.Json" Version="4.6.0" />
  </ItemGroup>
  <ItemGroup>
    <None Update="host.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
    <None Update="local.settings.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <CopyToPublishDirectory>Never</CopyToPublishDirectory>
    </None>
  </ItemGroup>
</Project>
```

**Comment by brettsam** [comment]
Okay, at least that narrows it down. Glad that the other package is working for you. We've made the change to get System.Data.SqlClient working as well -- that will roll out with our next host release within a few weeks.

**Comment by Robzilla** [comment]
Still getting the MIcrosoft.Data.SqlClient error even on the latest version (updated 8th Dec 2019).

Here is the cs.proj file and I am running Microsoft Visual Studio Enterprise 2019 Preview
Version 16.5.0 Preview 1.0.

My project is blocked on this. Connecting to SQL Server shouldn't be this hard. Can anyone advise how to collect extra debug to find out why the wrong dll is being copied/cached in the runtime?

```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>
    <AzureFunctionsVersion>v3-preview</AzureFunctionsVersion>
    <AssemblyName>TestSqlDb.Core</AssemblyName>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.Azure.Cosmos.Table" Version="1.0.5" />
    <PackageReference Include="Microsoft.Azure.Functions.Extensions" Version="1.0.0" />
    <PackageReference Include="Microsoft.Data.SqlClient" Version="1.1.0" />
    <PackageReference Include="Microsoft.NET.Sdk.Functions" Version="3.0.1" />
    <PackageReference Include="Microsoft.NETCore.App" Version="2.2.8" />
  </ItemGroup>
  <ItemGroup>
    <ProjectReference Include="..\LinkLibrary\LinkLibrary.csproj" />
  </ItemGroup>
  <ItemGroup>
    <None Update="host.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
    <None Update="local.settings.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <CopyToPublishDirectory>Never</CopyToPublishDirectory>
    </None>
  </ItemGroup>
</Project>
```

**Comment by brettsam** [comment]
Did you try targeting netcoreapp3.0? We don't yet support 3.1.

**Comment by brettsam** [comment]
Hmm, that may not be the issue. This is failing locally when you hit F5 from within VS? I just tried to reproduce this (even with netcoreapp3.1) and it worked fine for me locally, so there must be something else subtle happening.

Can you share your sample project on github somewhere and I can see if I can reproduce it locally?

**Comment by Robzilla** [comment]
According to the SqlClient team (6 days ago) I should upgrade to the latest version of .Net core (https://github.com/dotnet/SqlClient/issues/289#issuecomment-560128834). So I am a bit confused as to what version of .Net Core I should be using.

 I will create a simple project and see if I can just zip up and attach. But it obviously won't have the local settings with the SQL Server connection string.

**Comment by Robzilla** [comment]
Did a fresh install of VS2019 on a different machine and set the AzureFunctionsHiddenTagsVisible environment variable to 1 to get the new templates. But I am getting an object reference exception when using the "Azure Functions v3 Prereleasse (.NET Core)" project template. Are there any other tools/sdks that need to be installed?

**Comment by brettsam** [comment]
Is it like this? https://twitter.com/pksorensen/status/1202638801379696641. What version of VS 2019 are you using?

**Comment by Robzilla** [comment]
No its the generic message, have attached a screenshot. I am running a clean install of VS2019 Enterprise Preview (see details below).

Microsoft Visual Studio Enterprise 2019 Preview
Version 16.5.0 Preview 1.0
VisualStudio.16.Preview/16.5.0-pre.1.0+29521.150
Microsoft .NET Framework
Version 4.8.03752

Installed Version: Enterprise

Visual C++ 2019   00433-90100-95975-AA508
Microsoft Visual C++ 2019

ADL Tools Service Provider   1.0
This package contains services used by Data Lake tools

ASP.NET and Web Tools 2019   16.5.126.18061
ASP.NET and Web Tools 2019

ASP.NET Web Frameworks and Tools 2019   16.5.126.18061
For additional information, visit https://www.asp.net/

Azure App Service Tools v3.0.0   16.5.126.18061
Azure App Service Tools v3.0.0

Azure Data Lake Node   1.0
This package contains the Data Lake integration nodes for Server Explorer.

Azure Data Lake Tools for Visual Studio   2.4.1000.0
Microsoft Azure Data Lake Tools for Visual Studio

Azure Functions and Web Jobs Tools   16.5.126.18061
Azure Functions and Web Jobs Tools

Azure Stream Analytics Tools for Visual Studio   2.4.1000.0
Microsoft Azure Stream Analytics Tools for Visual Studio

C# Tools   3.5.0-beta1-19571-01+fb3f812a3e4b7534bef784fa2df1c21d1f67864d
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Fabric.DiagnosticEvents   1.0
Fabric Diagnostic Events

GitHub.VisualStudio   2.10.8.8121
A Visual Studio Extension that brings the GitHub Flow into Visual Studio.

IntelliCode Extension   1.0
IntelliCode Visual Studio Extension Detailed Info

Microsoft Azure HDInsight Azure Node   2.4.1000.0
HDInsight Node under Azure Node

Microsoft Azure Hive Query Language Service   2.4.1000.0
Language service for Hive query

Microsoft Azure Service Fabric Tools for Visual Studio   16.0
Microsoft Azure Service Fabric Tools for Visual Studio

Microsoft Azure Stream Analytics Language Service   2.4.1000.0
Language service for Azure Stream Analytics

Microsoft Azure Stream Analytics Node   1.0
Azure Stream Analytics Node under Azure Node

Microsoft Azure Tools   2.9
Microsoft Azure Tools for Microsoft Visual Studio 2019 - v2.9.21108.1

Microsoft Continuous Delivery Tools for Visual Studio   0.4
Simplifying the configuration of Azure DevOps pipelines from within the Visual Studio IDE.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft Library Manager   2.1.3+g9c91cdec24
Install client-side libraries easily to any web project

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual C++ Wizards   1.0
Microsoft Visual C++ Wizards

Microsoft Visual Studio Tools for Containers   1.1
Develop, run, validate your ASP.NET Core applications in the target environment. F5 your application directly into a container with debugging, or CTRL + F5 to edit & refresh your app without having to rebuild the container.

Microsoft Visual Studio VC Package   1.0
Microsoft Visual Studio VC Package

NuGet Package Manager   5.5.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit https://docs.nuget.org/

ProjectServicesPackage Extension   1.0
ProjectServicesPackage Visual Studio Extension Detailed Info

Snapshot Debugging Extension   1.0
Snapshot Debugging Visual Studio Extension Detailed Info

SQL Server Data Tools   16.0.61911.21070
Microsoft SQL Server Data Tools

ToolWindowHostedEditor   1.0
Hosting json editor into a tool window

TypeScript Tools   16.0.11113.2001
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   3.5.0-beta1-19571-01+fb3f812a3e4b7534bef784fa2df1c21d1f67864d
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual F# Tools 10.4 for F# 4.6   16.5.0-beta.19556.3+c2b6d5980a7f64dd4e46448b9b4f5d96eb415aeb
Microsoft Visual F# Tools 10.4 for F# 4.6

Visual Studio Code Debug Adapter Host Package   1.0
Interop layer for hosting Visual Studio Code debug adapters in Visual Studio

Visual Studio Container Tools Extensions (Preview)   1.0
View, manage, and diagnose containers within Visual Studio.

Visual Studio Tools for Containers   1.0
Visual Studio Tools for Containers

Visual Studio Tools for Kubernetes   1.0
Visual Studio Tools for Kubernetes

<img width="945" alt="VS2019-function-error" src="https://user-images.githubusercontent.com/3105346/70395917-88ea0c80-1a4f-11ea-9c4e-e82f4ca04e58.PNG">

**Comment by joefeser** [comment]
Same issue with Microsoft.NET.Sdk.Functions v3.0.1 and System.Data.SqlClient v4.8.0. Once I downgraded to v4.6.1, the problem went away.

**Comment by Robzilla** [comment]
Seriously it shouldn't be this hard to connect to SQL Server. This should be a core use case and tested thoroughly. Seems like quite a few people are having problems with this.

**Comment by brettsam** [comment]
@Robzilla / @joefeser  -- Can either of you create a repro on github that I can try? 

@Robzilla, I tried what you have above and I don't get the exception. I don't have your code or your referenced project -- so maybe something there is breaking it? But when I use our repro above (`var conn = new SqlConnection();`) -- it's able to load the type and continue. This failed a couple of versions ago b/c we weren't correctly copying native references in .NET Core 3 projects. Is there a different line of code throwing this for you? If so, I can try that.

If you can't get a repro together, you can try building from the command line with "dotnet build {yourproject} /bl" and emailing me your msbuild.binlog file -- my microsoft email is in my github profile.

**Comment by zpertee** [comment]
@brettsam FWIW, I've also tested what's above without issue.  The only issues I've had to date were when this was still in preview (i.e. issues expected) and those were quickly worked through-- thank you.  I agree that whatever @Robzilla issue is, it's likely confined to the machine and/or the specific code.

**Comment by erick2red** [comment]
Why is this issue closed? I can confirm it doesn't work yet. I just tried today with everything updated, and no, doesn't work.

**Comment by brettsam** [comment]
It's closed because the original issue above was fixed (as several people have confirmed).

It's clear that there are other scenarios broken here -- but I haven't gotten a repro that I've been able to run locally. 

@erick2red -- Can you share your project file with us so we can take a look? Ideally it'd be an entire project on github that fails for you so I can clone it and give it a try.

**Comment by erick2red** [comment]
My project files are these, first the function
```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>
    <AzureFunctionsVersion>v3</AzureFunctionsVersion>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.NET.Sdk.Functions" Version="3.0.1" />
    <PackageReference Include="Newtonsoft.Json" Version="12.0.3" />
    <PackageReference Include="StackExchange.Redis" Version="2.0.601" />
    <PackageReference Include="System.Text.Json" Version="4.7.0" />
  </ItemGroup>
  <ItemGroup>
    <ProjectReference Include="..\..\Libraries\CoreDbAccess\CoreDbAccess.csproj" />
  </ItemGroup>
  <ItemGroup>
    <None Update="host.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
    <None Update="local.settings.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <CopyToPublishDirectory>Never</CopyToPublishDirectory>
    </None>
  </ItemGroup>
</Project>
```

Then the dependency
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup Label="Globals">
    <SccProjectName>SAK</SccProjectName>
    <SccProvider>SAK</SccProvider>
    <SccAuxPath>SAK</SccAuxPath>
    <SccLocalPath>SAK</SccLocalPath>
    <Configurations>Debug;Release;ProductionPublish</Configurations>
    <Platforms>AnyCPU;x86;x64</Platforms>
  </PropertyGroup>

  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
    <TargetLatestRuntimePatch>true</TargetLatestRuntimePatch>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Dapper" Version="2.0.30" />
    <PackageReference Include="Dapper.Contrib" Version="2.0.30" />
    <PackageReference Include="Microsoft.CSharp" Version="4.7.0" />
    <PackageReference Include="StackExchange.Redis" Version="2.0.601" />
    <PackageReference Include="System.Data.SqlClient" Version="4.8.0" />
    <PackageReference Include="System.Text.Json" Version="4.7.0" />
    <PackageReference Include="System.Threading.Tasks.Extensions" Version="4.5.3" />
  </ItemGroup>

</Project>

```

**Comment by brettsam** [comment]
@erick2red -- It looks like your problem is with `System.Data.SqlClient`, which is a different underlying issue than `Microsoft.Data.SqlClient`. We've been discussing both here, which is getting confusing (I originally merged issues here b/c both assemblies had the same initial root cause -- so that's my fault for the confusion). 

However, it does look like we have a runtime issue here where we fixed this in v2 and the change did not make it to v3. Can you try a temporary workaround while we port the fix? Locally in your `local.settings.json`, add `"FUNCTIONS_V2_COMPATIBILITY_MODE": "true"`. And add the same App Setting in Azure. That will allow you to load the version of the assembly that you're referencing and I *think* that should unblock you for now. If so, let me know, and I'll update this when we've deployed the fix.

@Robzilla is still having problems with `Microsoft.Data.SqlClient` -- so if anyone has a repro with that, I'll continue investigating.

**Comment by Robzilla** [comment]
Hi @brettsam, any feedback on the getting VS2019 Preview working with the new function templates? It's fairly easy to reproduce. Just install a clean copy of VS2019 preview, turn on the new templates and try to create a new project using the new function templates. I have posted all the Visual Studio version info previously.

**Comment by brettsam** [comment]
@Robzilla -- I just installed a fresh VS2019 Preview, selected "ASP.NET and web development", "Azure development", and ".NET Core cross-platform development" and everything works. What options did you select?

@soninaren -- It looks like there may be some combo of VS options that causes the error above. Can you loop in someone from VS to see if they know how that can happen?

**Comment by erick2red** [comment]
@brettsam 

> Can you try a temporary workaround while we port the fix? Locally in your local.settings.json, add "FUNCTIONS_V2_COMPATIBILITY_MODE": "true". And add the same App Setting in Azure. That will allow you to load the version of the assembly that you're referencing and I think that should unblock you for now. If so, let me know, and I'll update this when we've deployed the fix.

Yes, this made it work. Thanks!

**Comment by DrBarnabus** [comment]
Just got hit by this and `FUNCTIONS_V2_COMPATIBILITY_MODE` seems to have fixed it for the time being.

Is this thread going to be updated when the fix is released so I can remove this?

**Comment by bmeijwaard** [comment]
Getting the same, tried some suggested stuff in this thread. 
I did get Azure Function to work with `SqlConnection` targetting Microsoft.Data.SqlClient 1.1.0, but as soon as a also target EntityFrameworksCore it fails again, also after upgrading to function v3 and netcoreapp3.1. What is found is that EFCore is targetting Microsoft.Data.SqlClient v1.0.19269.1 as i found in some project.assets.json

```
"Microsoft.EntityFrameworkCore.SqlServer/3.1.0": {
        "type": "package",
        "dependencies": {
          "Microsoft.Data.SqlClient": "1.0.19269.1",
          "Microsoft.EntityFrameworkCore.Relational": "3.1.0"
        },
        "compile": {
          "lib/netstandard2.0/Microsoft.EntityFrameworkCore.SqlServer.dll": {}
        },
        "runtime": {
          "lib/netstandard2.0/Microsoft.EntityFrameworkCore.SqlServer.dll": {}
        }
      },
```

**Comment by Ponomareff** [comment]
The message is displayed if **function.deps.json** is missing in **bin** folder, either locally or in Azure.

The problem is caused by invalid expression in **_FunctionsPostBuildDepsCopy** target
(in ...packages\microsoft.net.sdk.functions\3.0.1\build\Microsoft.NET.Sdk.Functions.Build.targets)
```
<Target Name="_FunctionsPostBuildDepsCopy" AfterTargets="PostBuildEvent" Condition="'$(SkipFunctionsDepsCopy)' != 'true'">
    <Copy SourceFiles="$(OutDir)$(ProjectName).deps.json" DestinationFiles="$(OutDir)bin\function.deps.json" Condition="Exists('$(OutDir)$(ProjectName).deps.json')"/>
  </Target>
```
It relies on **$(ProjectName)** which is not correct. It should be **$(AssemblyName)**. You are not able to reproduce the issue on 'Hello World' application, because both MSBuild variables have the same value.
Try to change assembly name to be different from Project name and put this line in your handler code 
```
var parser = new SqlConnectionStringBuilder("Data Source=(localdb)\\MSSQLLocalDB;Initial Catalog=master;Integrated Security=True;Connect Timeout=30;Encrypt=False;");
```
Hope this covers some scenarios listed here, @brettsam

**Comment by brettsam** [comment]
Great catch @Ponomareff. @fabiocav -- FYI.

**Comment by Robzilla** [comment]
@Ponomareff can you share your proj file and any other relevant config. At the moment I have the SQL Client people, pointing the finger at the Azure Function people who then point the finger that the Visual Studio people. All the product teams have their simple examples working and say "job done, works on my machine". Unfortunately the real world doesn't work that way :(

**Comment by fabiocav** [comment]
> All the product teams have their simple examples working and say "job done, works on my machine".

@Robzilla if you look at the interactions above, you'll see that isn't the case. It's unfortunate this is how you're perceiving this, but particularly in this area, fixes and improvements are validated based on known scenarios and what customers report, and those do become part of the validation suite.

@brettsam just merged a fix and we'll have that out ASAP, but if you can share the details of what is impacting you (i.e. share a repro for your case since you believe it isn't capture by what is currently validated against), we can ensure that addresses your problem as well. We'll update the issue as soon as the update is available so you can check if it helps you.

**Comment by brettsam** [comment]
I've just released to nuget Microsoft.Net.Sdk.Functions 1.0.31 (for v2 apps) and 3.0.2 (for v3 apps) that contain the fix from @Ponomareff above. For details on the releases see https://github.com/Azure/azure-functions-vs-build-sdk/releases.

There is still one known issue (and possibly more?) that require a runtime update in production:
- For v2, there is no workaround. But the fix is committed and the release should complete within a week. I will update back here when it is completed.
- For v3, you (interestingly enough...) set the `FUNCTIONS_V2_COMPATIBILITY_MODE` setting as detailed [above](https://github.com/Azure/Azure-Functions/issues/1370#issuecomment-563508311). _Side note -- this is because our v3 release was pushed out with newer bits than v2 -- and v2 doesn't actually have this fix yet._ I will update back here when this flag is no longer needed.

If anyone has more scenarios, please let me know the details and I can dig in.

**Comment by Robzilla** [comment]
@fabiocav I tried to create a example repo with the latest version of Visual Studio using the latest project templates but this failed with an exception (https://github.com/Azure/Azure-Functions/issues/1370#issuecomment-562991756). So I couldn't even create a repo to share and nobody seems to be able to advise. I would have thought being able to create a new Azure functions project with up to date tools and templates should at least work out of the box. But it doesn't seem that way and I am getting silence from MS on this.

**Comment by brettsam** [comment]
@Robzilla -- I attempted to reproduce this and told you what I chose for my installation options [here](https://github.com/Azure/Azure-Functions/issues/1370#issuecomment-564045541). The list you've provided can be helpful but I'm not sure how to reproduce that list given he Workload selection UX that the Installer provides. But since this is unrelated to the issue here and this is getting off track, I've created https://github.com/Azure/Azure-Functions/issues/1434 to narrow that down. Please hop over there and we can try to figure out the issue, looping in the VS experts if we need them.

**Comment by bmeijwaard** [comment]
Just updated Functions SDK to 3.0.2. Still got `(Microsoft.Data.SqlClient is not supported on this platform.)`
Currently running on netcoreapp3.1 and functions v3
Also set `"FUNCTIONS_V2_COMPATIBILITY_MODE": "true"` in `local.settings.json`

For futher investigation I could provide my entire repo.

**Comment by fabiocav** [comment]
@bmeijwaard if you could share a repository  with a repro, that would indeed help.

**Comment by bmeijwaard** [comment]
> @bmeijwaard if you could share a repository with a repro, that would indeed help.

Is there a way to send you the 'zipped' code in private, since it is a private repo?

After looking into it a little deeper I suspect that the issue is coming from `Microsoft.AspNetCore.Identity.EntityFrameworkCore`
in https://github.com/aspnet/AspNetCore/blob/master/eng/Versions.props 
`line 264` version 1.0.19249.1 of Microsoft.Data.SqlClient is referenced. In one of my project depedencies I have referenced Microsoft.AspNetCore.Identity.EntityFrameworkCore v3.1.0 using IdentityDbContext class.

**Comment by brettsam** [comment]
@bmeijwaard -- If you want to contact me via email (my microsoft mail is in my github profile), we can work through it that way, then come back here with any findings. We likely don't need your code -- just the project structure with project files.

**Comment by bmeijwaard** [comment]
Awesome, ill forward the mail I sent to Fabio.

**Comment by Ponomareff** [comment]
@bmeijwaard Before you struggle with reproduction steps, please verify the following for your project:
**function.deps.json** and **runtimes** are both present in **bin** folder.
![image](https://user-images.githubusercontent.com/5252051/70804871-8bb56c00-1d85-11ea-85d7-612fc7140afd.png)
Same for your deployed Azure Function:
![image](https://user-images.githubusercontent.com/5252051/70805388-cec40f00-1d86-11ea-8dea-9a7fcb0b4453.png)

**Comment by bmeijwaard** [comment]
@Ponomareff this is the bin folder of the project.
![image](https://user-images.githubusercontent.com/31803222/70806175-ff25a080-1dba-11ea-82c2-b93cfb4765ec.png)

**Comment by lluppes3M** [comment]
I just updated a function to .NET Core 3.1 today and had this exact same issue, which I was able to resolve based on this post.  I added the setting "FUNCTIONS_V2_COMPATIBILITY_MODE": "true" to the local.settings and the server function and I updated my project file to copy the function.deps.json and that resolved the issue.   Up until I did that, it worked fine locally but generated the error "Microsoft.Data.SqlClient is not supported on this platform." on the server

My .csProj file:
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>
    <AzureFunctionsVersion>v3</AzureFunctionsVersion>
    <AssemblyName>ProcessAlerts</AssemblyName>
    <RootNamespace>ProcessAlerts</RootNamespace>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.Azure.WebJobs.Extensions.CosmosDB" Version="3.0.5" />
    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="3.1.0" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Relational" Version="3.1.0" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="3.1.0" />
    <PackageReference Include="Microsoft.NET.Sdk.Functions" Version="3.0.2" />
  </ItemGroup>
  <ItemGroup>
    <None Update="host.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
    <None Update="local.settings.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <CopyToPublishDirectory>Never</CopyToPublishDirectory>
    </None>
  </ItemGroup>
	<Target Name="PostPublish" BeforeTargets="Publish">
	  <Exec Command="copy $(PublishDir)$(ProjectName).deps.json $(PublishDir)bin\function.deps.json" />
	</Target>
</Project>

**Comment by Eirenarch** [comment]
Hit the System.Data.SqlCLient version of the problem today with the 3.0 (non-preview) release. The FUNCTIONS_V2_COMPATIBILITY_MODE did fix it but I am still wondering if I will ever be able to remove it after an update

**Comment by Robzilla** [comment]
@brettsam well I did get this running locally but failed on the server even after upgrading to the latest libs. See proj file below. I can confirm there is no function.deps.json file in the bin dir on the server as per https://github.com/Azure/Azure-Functions/issues/1370#issuecomment-565450516
Hope that narrows it down. Any chance this is fixed in version 3.0.3?

```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>
    <AzureFunctionsVersion>v3</AzureFunctionsVersion>
    <AssemblyName>TestDb.Functions</AssemblyName>
    <RootNamespace>TestDb.Functions</RootNamespace>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.Azure.Cosmos.Table" Version="1.0.6" />
    <PackageReference Include="Microsoft.Data.SqlClient" Version="1.1.0" />
    <PackageReference Include="Microsoft.NET.Sdk.Functions" Version="3.0.2" />
  </ItemGroup>
  <ItemGroup>
    <None Update="host.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
    <None Update="local.settings.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <CopyToPublishDirectory>Never</CopyToPublishDirectory>
    </None>
  </ItemGroup>
</Project>
```

**Comment by soninaren** [comment]
@Robzilla if you are seeing the `function.deps.json` in the bin folder locally, but are not seeing the same on the server. Then i would recommed using the `Run from Package file` for deployment. The out logs from VS should show where the package (zip file) was generated locally. The contents of the function app on the server should match exactly with the contents of the zip file. Hope this helps.

**Comment by Robzilla** [comment]
@soninaren and @brettsam I am publishing to an existing app service plan via the zip deploy method in Visual Studio 2019. I am also creating a debug build and have set Delete Existing Files to true.

I have checked the zip file that is produced during the publish and function.deps.json is not there.

**Comment by fabiocav** [comment]
@Robzilla just wanted to give you an update. Things are a bit slower as some folks are out of the office for the holidays, but I'm blocking some time to look at the repro you've provided this week. Will provide an update as soon as we have more information.

If the deps file is not in the deployment payload, that would definitely explain why you're running into the problem, but would require a bit more investigation.

**Comment by Robzilla** [comment]
@fabiocav just let me know what debug you require from Visual Studio.

**Comment by fabiocav** [comment]
@Robzilla we have just made an SDK fix to address the issue you're seeing. We'll have that deployed at the beginning of the year, but we could make it available to you on a private feed before then, if you believe that would be helpful

@bmeijwaard thank you again for providing a repro for this, as the issue was triggered by a specific setup you have with your solution (shared more information over email). We have identified an issue here and will have the fix for this rolling out at the beginning of the year as well.

/cc @brettsam

**Comment by eastmane** [comment]
I seem to have hit a similar issue (I think) to @bmeijwaard  - using EF Core, VS 16.4.2, and functions runtime 3.0.12930.0 - I get "Microsoft.Data.SqlClient is not supported on this platform" with or without FUNCTIONS_V2_COMPATIBILITY_MODE. I'm referencing Microsoft.Data.SqlClient 1.1.0 everywhere I can find it and the functions.deps.json and runtimes folder are present in bin.

I noticed that the file size on the Microsoft.Data.SqlClient in the bin folder is different to the filesize in my .nuget and the runtimes version - inspired by https://stackoverflow.com/a/53952815/908 I included this in .csproj:

``` 
  <!-- For publish -->
  <ItemGroup>
    <None Include="$(USERPROFILE)\.nuget\packages\microsoft.data.sqlclient\1.1.0\runtimes\win\lib\netcoreapp2.1\Microsoft.Data.SqlClient.dll">
      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
    </None>
  </ItemGroup>
  <!-- For local debug -->
  <Target Name="CopyToBin" BeforeTargets="Build">
    <Copy SourceFiles="$(USERPROFILE)\.nuget\packages\microsoft.data.sqlclient\1.1.0\runtimes\win\lib\netcoreapp2.1\Microsoft.Data.SqlClient.dll" DestinationFolder="$(OutputPath)\bin" />
  </Target>
```

This copies the .nuget version into bin and it seems to work fine.

Any ideas what's going on here and why I'm getting a seemingly incompatible dll included by default?

Happy to share projects etc if it'd help.

**Comment by fabiocav** [comment]
@eastmane the best option would really be to open a separate issue and share the repro if you could. Looking at these kinds of issues without more context is tricky as large number of problems may end up manifesting themselves with similar failures, which tends to be misleading. A good example of that was the issue @bmeijwaard ran into, which had the same exception, but it was caused by a parsing and mapping issue recently introduced in the runtime.

**Comment by eastmane** [comment]
Thanks for your help - have opened a separate issue, I can't post repro on here as it's part of a larger project but can probably share privately.

**Comment by fabiocav** [comment]
Thank you. Will continue following up on the newly created issue. For the repro, I'd be happy to look at something shared privately as well. My email is available on my GitHub profile.

**Comment by metoule** [comment]
> The problem is caused by invalid expression in **_FunctionsPostBuildDepsCopy** target
> (in ...packages\microsoft.net.sdk.functions\3.0.1\build\Microsoft.NET.Sdk.Functions.Build.targets)
> 
> ```
> <Target Name="_FunctionsPostBuildDepsCopy" AfterTargets="PostBuildEvent" Condition="'$(SkipFunctionsDepsCopy)' != 'true'">
>     <Copy SourceFiles="$(OutDir)$(ProjectName).deps.json" DestinationFiles="$(OutDir)bin\function.deps.json" Condition="Exists('$(OutDir)$(ProjectName).deps.json')"/>
>   </Target>
> ```
> 
> It relies on **$(ProjectName)** which is not correct. It should be **$(AssemblyName)**. You are not able to reproduce the issue on 'Hello World' application, because both MSBuild variables have the same value.

The same issue is present in [Microsoft.NET.Sdk.Functions.Publish.targets](https://github.com/Azure/azure-functions-vs-build-sdk/blob/35c4d4e448877495a2e25e31166b6e4e3fe47eb5/src/Microsoft.NET.Sdk.Functions.MSBuild/Targets/Microsoft.NET.Sdk.Functions.Publish.targets#L120): 

    <Target Name="_FunctionsPublishDepsCopy" BeforeTargets="Publish" Condition="'$(SkipFunctionsDepsCopy)' != 'true'">
        <Copy SourceFiles="$(PublishDir)$(ProjectName).deps.json" DestinationFiles="$(PublishDir)bin\function.deps.json" Condition="Exists('$(PublishDir)$(ProjectName).deps.json')"/>
    </Target>

which causes an issue when deploying on Azure.

**Comment by fabiocav** [comment]
@metoule that issue has been addressed and the package is pending deployment. If you'd like to validate, you can consume the package from here in the meantime: https://www.myget.org/feed/azure-appservice-staging/package/nuget/Microsoft.NET.Sdk.Functions/3.0.3

/cc @brettsam

**Comment by lars-erik** [comment]
@fabiocav I just resolved this by installing 3.0.3 from staging. Looks good! 🙂👍

**Comment by ldfloresgon** [comment]
@fabiocav  it seems fixed with Microsoft.NET.Sdk.Functions -Version 3.0.2 isn't it?

**Comment by curelom** [comment]
Negative.  I've installed Microsoft.NET.Sdk.Functions Version 3.0.2 and it still doesn't work.  This issue should not have been closed.  Please re-open

**Comment by ldfloresgon** [comment]
@curelom you must clean and rebuild the solution and will work fine

**Comment by curelom** [comment]
No good.  Cleaned the solution, even deleted bin and obj folders.  Still get System.Data.SqlClient is not supported on this platform.

**Comment by ldfloresgon** [comment]
@curelom  you should use microsoft.data.sqlclient instead of system.data.sqlclient

**Comment by curelom** [comment]
@ldfloresgon Thank you for that catch.  I had referenced that as SqlParameter in one class.  I have removed that now and get the  System.PlatformNotSupportedException: Microsoft.Data.SqlClient is not supported on this platform error

**Comment by mjhoffmeister** [comment]
I was able to get everything working after switching to Microsoft.Data.SqlClient, then uninstalling System.Data.SqlClient and doing a clean/rebuild. I initially upgraded to Microsoft.NET.Sdk.Functions 3.0.3, but I just reverted to 3.0.2; everything is still working.

**Comment by OwenPattison** [comment]
Im using .Net Core 2.2 and I have updated the Microsoft.NET.Sdk.Functions from 1.0.29 to 1.0.31 and I also see Microsoft.Data.SqlClient is not supported on this platform error.

Reverting the package to 1.0.29 solves the problem for me while im using .Net Core 2.2 however from what I can see this is not listed as a breaking change which I feel it should be.

**Comment by bestcoder** [comment]
> @metoule that issue has been addressed and the package is pending deployment. If you'd like to validate, you can consume the package from here in the meantime: https://www.myget.org/feed/azure-appservice-staging/package/nuget/Microsoft.NET.Sdk.Functions/3.0.3
> 
> /cc @brettsam

FWIW, I tried the 3.03 package in my azure function project that is getting the error and there was no change. The error still occurs. I also tried the Microsoft.NET.Sdk.Functions.3.0.4 package and the error still occurs with that. My project is targeting '.NETFramework,Version=v4.8'

**Comment by daunish** [comment]
I am still seeing this issue also. I am migrating from Core 2.2 to Core 3.1
I've upgraded all packages to the latest - the functions sdk is a 3.0.3
I set the functions runtime to v3 in my csproj file
The project is targeting .NETStandard 2.1

I see this when I try to execute a function:

`2020-02-22T09:47:30.676 [Error] Executed '<FUNCTION-NAME>' (Failed, Id=2423fc5a-397a-44f6-a6d2-9b6e2636ce1e)
Microsoft.Data.SqlClient is not supported on this platform.`

What's a bit concerning here is that it is trying to use the correct client (`Microsoft.Data.SqlClient`) but it's not working. Would love some guidance on how to get around this...

**Comment by daunish** [comment]
I was able to get this working by downgrading the functions sdk to 3.0.2 (which is odd considering the fix was supposed to go to into 3.0.3)
And I changed the target to `netcoreapp3.0` - when I tried `netcoreapp3.1` it did not work.

**Comment by jadenrogers** [comment]
@daunish had same issue and is now working on 3.0.5 - netcoreapp3.0

**Comment by c-lamont** [comment]
Hi, I've been trying all the tips in this thread so far for a whole day but still getting `Microsoft.Data.SqlClient is not supported on this platform.`

It has been working for the past couple of years and only when I updated my .Net Core version did this error start appearing (I believe). Here are my installed .Net Core versions.
```
2.1.202 [/usr/local/share/dotnet/sdk]
3.1.100 [/usr/local/share/dotnet/sdk]
3.1.101 [/usr/local/share/dotnet/sdk]
3.1.102 [/usr/local/share/dotnet/sdk]
3.1.200 [/usr/local/share/dotnet/sdk]
```

Microsoft.NET.sdk.functions: 3.0.6 (Also tried 3.0.2 as suggested)
Microsoft.Data.SqlClient: 1.1.2
MacOS: 10.15.4
Visual Studio Mac: 8.5.3

Any further advice would be greatly appreciated.

**Comment by WinInsider** [comment]
@brettsam @fabiocav

Is it possible to overcome "Microsoft.Data.SqlClient is not supported on this platform" exception **without** using "Microsoft.NET.Sdk.Functions" nuget package?

As it has **a lot** of dependencies (heavy)…
To overcome the "not supported" exception, is it enough to:
1) rename "runtimes" to "bin" folder
2) copy ProjectName.deps.json to "bin"
3) have function.json in sub-folder

Targeting netcoreapp3.1 and Function Runtime 3

**Comment by paulbatum** [comment]
Just a quick note - we consider the `Microsoft.NET.Sdk.Functions` NuGet package as a required part of the .NET developer experience for Azure Functions. We do not test or support scenarios where this package has been removed.

**Comment by WinInsider** [comment]
@paulbatum `Microsoft.NET.Sdk.Functions` NuGet package pulls A LOT dependencies. Not every consumer will make use of them all... For example, AspNetCore is pulled (indirectly via Microsoft.Azure.WebJobs.Extensions.Http). Azure Queue scenarios has no need for AspNetCore.

Consider separating MSBuld related task into separate package.  Light, portably and not tying to any particular stack is very appealing to a lot of consumers.  Regardless, focus on developer experience and integration is what set Azure apart from others !!! And makes it easy to use !!!

**Comment by paulbatum** [comment]
@WinInsider Yes what you're asking for makes sense, there is definite room for improvement in terms of the minimal dependency surface area when writing a .NET based function. Anyway I don't want to derail this thread further. 

For the folks still reporting problems when running on the latest package (3.0.7 at time of writing) and targetting netcoreapp3.1, please file a new issue and link back to this one. Fill out the issue template and make sure you include additional relevant details such as the contents of your csproj.

**Comment by npnelson** [comment]
I have the problem with the 3.0.7 runtime, although I can only reproduce the problem when running in a container. Non container locally and in Azure seems to work fine, but I love containers, so it would be nice to get it working there too.

I commented with a github repo that can reproduce it when running  with Visual Studio on this newer issue https://github.com/Azure/azure-functions-host/issues/5950

**Comment by npnelson** [comment]
Interesting enough, I just tried within Visual Studio as a Release build and it worked fine. I am sure why Release works and Debug doesn't. I was also able to successfully to successfully build the container from Azure Pipelines and run it in an Azure App Service Linux container instance. 

I'm still not sure why it doesn't work for me with a Visual Studio debug build, but that isn't a big issue for me now since I can run it in production.

**Comment by bfmsoft** [comment]
I just confirmed that is in "Microsoft.NET.Sdk.Functions" Version="3.0.8".  Move back to 3.0.7 and goes away.

**Comment by TheHokieCoder** [comment]
> I just confirmed that is in "Microsoft.NET.Sdk.Functions" Version="3.0.8". Move back to 3.0.7 and goes away.

+1...been banging my head against my desk for several hours now trying to figure out why I was getting the platform not supported message when all indication was that it was working. After downgrading the SDK to v3.0.7, the SQL client is working.

**HOWEVER**, after digging around through some other things, I stumbled upon what may be a caveat or potentially a regression in the SDK tooling that I was able to work around and get the SqlClient package working with v3.0.8 of the SDK.

It appears to me that in v3.0.7, when you build your function app the build target places the `runtimes` directory under the `bin` directory (as mentioned earlier in this issue). The same happens when you publish the function app. But when you update the SDK to v3.0.8, the `runtimes` directory is left _alongside_ the `bin` directory...not _inside_ it...causing the exception. When you publish with v3.0.8, the `runtimes` directory is, in fact, moved into the `bin` directory and no exception occurs.

So with v3.0.8, my current workaround is to add a post-build target that is similar to what IGx89 mentioned in an earlier [comment](#issuecomment-546030332). Note, however, that the target is post-**build**, not post-publish. The difference with my approach is that the files are copied after build, which supports local debugging of your function app while also not stepping on the toes of what the publish target does. The downside is that there are multiple copies of runtime binaries in your build directories, but the upside is that this won't affect what is copied to the publish directory during that target.
```xml
  <Target Name="PostBuild" AfterTargets="AfterBuild">
    <Exec Command="xcopy /E /I /Q /Y $(OutDir)runtimes $(OutDir)bin\runtimes" />
  </Target>
```

The [azure-functions-vs-build-sdk](https://github.com/Azure/azure-functions-vs-build-sdk/) repository doesn't seem to be very well maintained in terms of tagging their releases and being transparent about what commits are going into which releases. As an example, see issue [#438](https://github.com/Azure/azure-functions-vs-build-sdk/issues/438). I wasn't able to spend enough time to follow what changed from v3.0.7 to v3.0.8...but my hunch is that there was some sort of regression.

**Comment by pietervdheijden** [comment]
> 
> 
> > I just confirmed that is in "Microsoft.NET.Sdk.Functions" Version="3.0.8". Move back to 3.0.7 and goes away.
> 
> +1...been banging my head against my desk for several hours now trying to figure out why I was getting the platform not supported message when all indication was that it was working. After downgrading the SDK to v3.0.7, the SQL client is working.
> 
> **HOWEVER**, after digging around through some other things, I stumbled upon what may be a caveat or potentially a regression in the SDK tooling that I was able to work around and get the SqlClient package working with v3.0.8 of the SDK.
> 
> It appears to me that in v3.0.7, when you build your function app the build target places the `runtimes` directory under the `bin` directory (as mentioned earlier in this issue). The same happens when you publish the function app. But when you update the SDK to v3.0.8, the `runtimes` directory is left _alongside_ the `bin` directory...not _inside_ it...causing the exception. When you publish with v3.0.8, the `runtimes` directory is, in fact, moved into the `bin` directory and no exception occurs.
> 
> So with v3.0.8, my current workaround is to add a post-build target that is similar to what IGx89 mentioned in an earlier [comment](#issuecomment-546030332). Note, however, that the target is post-**build**, not post-publish. The difference with my approach is that the files are copied after build, which supports local debugging of your function app while also not stepping on the toes of what the publish target does. The downside is that there are multiple copies of runtime binaries in your build directories, but the upside is that this won't affect what is copied to the publish directory during that target.
> 
> ```
>   <Target Name="PostBuild" AfterTargets="AfterBuild">
>     <Exec Command="xcopy /E /I /Q /Y $(OutDir)runtimes $(OutDir)bin\runtimes" />
>   </Target>
> ```
> 
> The [azure-functions-vs-build-sdk](https://github.com/Azure/azure-functions-vs-build-sdk/) repository doesn't seem to be very well maintained in terms of tagging their releases and being transparent about what commits are going into which releases. As an example, see issue [#438](https://github.com/Azure/azure-functions-vs-build-sdk/issues/438). I wasn't able to spend enough time to follow what changed from v3.0.7 to v3.0.8...but my hunch is that there was some sort of regression.

This worked for me! Thanks a ton :)

**Comment by flatproject** [comment]
Downgrading to 3.0.7 worked for me

**Comment by Gabsch** [comment]
> For the folks still reporting problems when running on the latest package (3.0.7 at time of writing) and targetting netcoreapp3.1, please file a new issue and link back to this one. Fill out the issue template and make sure you include additional relevant details such as the contents of your csproj.

@paulbatum Do we need a new issue for 3.0.8 or are you guys aware of the problem?

**Comment by bbakermmc** [comment]
Azure is such a shit show, they break stuff every minor patch and you spend hours or days trying to figure out wtf is going on.  Its like basically never upgrade your packages.

**Comment by brettsam** [comment]
This was an issue with the build tooling that came with 3.0.8 -- Issues and PR are here:
- https://github.com/Azure/azure-functions-vs-build-sdk/issues/436
- https://github.com/Azure/azure-functions-vs-build-sdk/issues/437
- https://github.com/Azure/azure-functions-vs-build-sdk/pull/443

**Comment by c-lamont** [comment]
My fix was changing my project from
```
<TargetFramework>netstandard2.0</TargetFramework>
<AzureFunctionsVersion>v2</AzureFunctionsVersion>
```
To
```
<TargetFramework>netcoreapp3.1</TargetFramework>
<AzureFunctionsVersion>v3</AzureFunctionsVersion>
```
Then updating package `Microsoft.NET.Sdk.Functions` to v 3.0.7

**Comment by alnintex** [comment]
Microsoft.NET.Sdk.Functions to v 3.0.9 fixed the problem

**Comment by TechWatching** [comment]
Can't upgrade to v3.0.9 because of this [issue](https://github.com/Azure/azure-functions-vs-build-sdk/issues/470) :(

**Comment by frasparacio** [comment]
Still having the same problem with Microsoft.NET.Sdk.Functions v. 3.0.9.
Just migrating from .net core 2.2 to 3.1 and this was one of the problems we had.
Solved putting postbuild and postpublish command as suggested above, thanks!

**Comment by ZooDoo4U** [comment]
So i just ran into the issue,  easy to understand if this is beta, but isn't everything with issues officially released at first it seems looking at the date of the original post this has been going on for over a year now?  Last comment only 10 hours ago...

**Comment by anthonychu** [comment]
@brettsam Do you know what could still be causing this?

@ZooDoo4U Can you solve it using the post build and post publish commands above?

**Comment by fabiocav** [comment]
@anthonychu to understand what is happening, we'd need the details.

@ZooDoo4U can you please open a new issue with details of all version used, exception stacks, etc? we can take a closer look then. 

Will lock this issue as there have been tooling and runtime updates to address the original root cause, so any new instances will need to be investigated separately.

**Comment by brettsam** [comment]
Please include your csproj in the repro as that helps determine what could be going on.

**Comment by cgillum** [comment]
If anyone runs into this while using the functions CLI to install extensions (`func extensions install`) that depend on Microsoft.Data.SqlClient, I was able to work around this by updating my **extensions.csproj** file to look like the following:

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>
    <WarningsAsErrors></WarningsAsErrors>
    <DefaultItemExcludes>**</DefaultItemExcludes>
  </PropertyGroup>

  <!-- Workaround for https://github.com/Azure/Azure-Functions/issues/1370 -->
  <Target Name="PostBuild" AfterTargets="PostBuildEvent">
    <Move SourceFiles="$(OutDir)/extensions.deps.json" DestinationFiles="$(OutDir)/function.deps.json" />
  </Target>

  <ItemGroup>
    <!-- ...(other dependencies)... -->
    <PackageReference Include="Microsoft.Data.SqlClient" Version="3.0.0" />
  </ItemGroup>
</Project>
```

This workaround specifically applies to C# scripting and non-.NET languages which aren't using extension bundles.
